### PR TITLE
Added k-Nearest neighbor search to RTree

### PR DESCRIPTION
**Pull request by schedul-xor**
I believe QuadTree can also have kNN searches too

### CODE REVIEW

File: vtm/src/org/oscim/utils/GeoPointUtils.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ public static boolean contains(GeoPoint[] geoPoints, GeoPoint geoPoint) {
     }
 
     /**
-     * Returns the distance between the given segment and point.
+     * Returns the squareDistance between the given segment and point.
```
Comment (devemux86) [inline vtm/src/org/oscim/utils/GeoPointUtils.java]: Is this after an automatic refactoring of the IDE?

Comment (schedul-xor) [reply vtm/src/org/oscim/utils/GeoPointUtils.java]: It seems so. I'll fix it

**Comment by devemux86** [comment]
As a general note, since it's adapted code from elsewhere, better if retain the methods and variable names, would be convenient for comparison of the integration.

**Comment by devemux86** [comment]
Also can include also some of the optional parameters, if that's doable as shown in https://github.com/mourner/rbush-knn?
e.g. `maxDistance`

**Comment by schedul-xor** [comment]
> As a general note, since it's adapted code from elsewhere, better if retain the methods and variable names, would be convenient for comparison of the integration.

I understand, but in this case some variable names used in original source code was misleading (for example the `minDistance` parameter was actually a *squared* minDistance), so I fixed by my own decision.

**Comment by devemux86** [comment]
If it's used for comparisons then it's essential for performance to avoid on-top calculations. And since it's internal can follow original author to have easy reference for the future readings.

File: vtm/src/org/oscim/utils/RTree.java [hunk 2]
```diff
@@ -355,6 +369,70 @@ public boolean search(Box bbox, SearchCb<T> cb, Object context) {
...
+    @Override
+    public List<T> searchKNearestNeighbors(Point center, int k, double maxDistance, List<T> results) {
+        // Refered https://github.com/mourner/rbush-knn/blob/master/index.js
+
+        if (results == null)
+            results = new ArrayList<T>(16);
+        results.clear();
+
+        PriorityQueue<KnnItem> queue = new PriorityQueue<>();
+        double maxSquareDistance = maxDistance * maxDistance;
```
Comment (devemux86) [inline vtm/src/org/oscim/utils/RTree.java]: Cannot see in original code such action.
And what happens if max is infinity?

Comment (schedul-xor) [reply vtm/src/org/oscim/utils/RTree.java]: The action is found in https://github.com/mourner/rbush-knn/pull/14 but seems to be not merged yet. In this current pull request `maxDistance * maxDistance` is done per children iteration, but it is clear that should be done only once before the iteration to improve performance.

`Double.POSITIVE_INFINITY * Double.POSITIVE_INFINITY` is `Double.POSITIVE_INFINITY`. Current unit tests are already using it.

**Comment by devemux86** [comment]
Thanks, merged via ffeaf1b81c50e1022b0888914639e93f3bd6da6d.

**Pull request closed without merging**